### ISSUE TITLE
select! cannot contain more than 64 macro calls

### ISSUE THREAD

**Issue by dbrgn** [rust-lang/futures-rs#2097]
This is a bit strange, but based on experimentation, I hope that the issue title is right.

The following code fails to compile:

```rust
#![recursion_limit = "2048"]

use futures::{select, channel::mpsc};
use futures_util::sink::SinkExt;
use futures_util::stream::StreamExt;

async fn blah() {
    let (mut tx, mut rx) = mpsc::channel(8);
    tx.send(42u8).await;
    
    loop {
        let mut next_msg = rx.next();
        select! {
            _ = next_msg => {
                println!("1");
                ...
                println!("65"); 
            }
        }
    }
}

fn main() {
    let myfuture = blah();
}
```

Error message:

```rust
   Compiling playground v0.0.1 (/playground)
error: no rules expected the token `!`
  --> src/main.rs:13:9
   |
13 | /         select! {
14 | |             _ = next_msg => {
15 | |                 println!("1");
16 | |                 println!("2");
...  |
80 | |             }
81 | |         }
   | |_________^ no rules expected this token in macro call
   |
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

Removing one of the println calls makes it compile.

The problem is not specific to `println!`, it also happens with macros like `debug!`.

Reproducer: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=de8c4b5aa1ae3ac9c432952e9143fb59

This happened to me in a real-world project due to logging calls in a big select block :confused:

**Comment by cramertj** [comment]
Presumably this is a limitation of proc-macro-hack? cc @dtolnay

**Comment by dtolnay** [comment]
Yeah this limit comes from https://github.com/dtolnay/proc-macro-hack/blob/0.5.11/nested/build.rs#L29. I guess we could bump it. I don't know how much it would slow down rustc's macro matcher; the generated macro is quadratically large in this limit.

**Comment by dbrgn** [comment]
If possible, at least a proper error msg should be generated. I managed to reduce the issue a bit by converting a commonly used custom macro to a regular function.

**Comment by dtolnay** [comment]
I fixed the error message in proc-macro-nested 0.1.4.

```console
error: this macro does not support >64 nested macro invocations
  --> src/main.rs:13:9
   |
13 | /         select! {
14 | |             _ = next_msg => {
15 | |                 println!("1");
16 | |                 println!("2");
...  |
80 | |             }
81 | |         }
   | |_________^
```

**Comment by taiki-e** [comment]
This will be fixed by https://github.com/dtolnay/proc-macro-hack/issues/62.
